charmcraft/commands/pack.py

```diff
@@ -0,0 +1,130 @@
+    zipfh.close()
+
+
+def get_paths_to_include(dirpath):
+    """Get all file/dir paths to include."""
+    allpaths = set()
+
+    # all mandatory files, which must exist (currently only bundles.yaml is mandatory, and
+    # it's verified before)
+    for fname in MANDATORY_FILES:
+        allpaths.add(dirpath / fname)
+
+    # the extra files, which must be relative
+    config = load_yaml(dirpath / 'charmcraft.yaml') or {}
+    prime_specs = config.get('parts', {}).get('bundle', {}).get('prime', [])
+
+    for spec in prime_specs:
+        if spec[0] == os.path.sep:
+            raise CommandError(
+                "Extra files in prime config can not be absolute: {!r}".format(spec))
+
+        fpaths = sorted(fpath for fpath in dirpath.glob(spec) if fpath.is_file())
+        logger.debug("Including per prime config %r: %s.", spec, fpaths)
+        allpaths.update(fpaths)
+
+    return sorted(allpaths)
+
+
+_overview = """
+Build the bundle, leaving a .zip file as the result of the process.
+
+You can `juju deploy` directly from the resulting .zip file, or upload it to
+the store (see the "upload" command).
+"""
+
+
+class PackCommand(BaseCommand):
+    """Build the bundle or the charm.
+
+    Eventually this command will also support charms, but for now it will work only
+    on bundles.
+    """
+    name = 'pack'
+    help_msg = "Build the bundle"
+    overview = _overview
+
+    def fill_parser(self, parser):
+        """Add own parameters to the general parser."""
+        parser.add_argument(
+            '-f', '--from', type=pathlib.Path, dest='from_dir',
+            help="The directory where the bundle project is located, where the build "
+                 "is done from; defaults to '.'")
+
+    def run(self, parsed_args):
+        """Run the command."""
+        if parsed_args.from_dir is None:
+            dirpath = pathlib.Path.cwd()
+        else:
+            dirpath = parsed_args.from_dir.expanduser()
+            if not dirpath.exists():
+                raise CommandError("Bundle project directory was not found: '{}'.".format(dirpath))
+            if not dirpath.is_dir():
+                raise CommandError(
+                    "Bundle project directory is not really a directory: '{}'.".format(dirpath))
+
+        # get the config files
+        bundle_filepath = dirpath / 'bundle.yaml'
+        bundle_config = load_yaml(bundle_filepath)
+        if bundle_config is None:
+            raise CommandError(
+                "Missing or invalid main bundle file: '{}'.".format(bundle_filepath))
+        bundle_name = bundle_config.get('name')
+        if not bundle_name:
+            raise CommandError(
+                "Invalid bundle config; missing a 'name' field indicating the bundle's name in "
+                "file '{}'.".format(bundle_filepath))
+
+        charmcraft_filepath = dirpath / 'charmcraft.yaml'
+        charmcraft_config = load_yaml(charmcraft_filepath)
+        if charmcraft_config is None:
```

Comment:
At some point we'll probably want to validate the schema of the config files.

Reply:
Good idea!